### PR TITLE
New WC_Data class with meta methods for CRUD

### PR DESCRIPTION
**Pull request by emdashcodes**
This PR replaces the WC_Data interface with an Abstract WC_Data class (the create, save, etc methods are now abstract classes that child classes implement). The class also introduces a set of meta helper methods (taken from the order branch, but updated to work with users) that all CRUD classes can use. This can also be a future home for any other shared logic. Includes tests and a mock inherited WC_Data class for testing meta methods.

### CODE REVIEW

File: includes/abstracts/abstract-wc-data.php [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+					$this->_meta_data[ $meta_id ] = (object) array(
+						'key'   => $meta['key'],
+						'value' => $meta['value'],
+					);
+				}
+			}
+		}
+	}
+
+
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-data.php]: extra newline here

File: includes/abstracts/abstract-wc-data.php [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+			$object_id_field = 'post_id';
+		} elseif ( 'user' === $this->_meta_type ) {
+			$object_id_field = 'user_id';
+			$meta_id_field   = 'umeta_id';
+		} else if ( 'comment' === $this->_meta_type ) {
+			$object_id_field = 'comment_id';
+		} else if ( 'term' === $this->_meta_data ) {
+			$object_id_field = 'term_id';
+		} else {
+			if ( ! empty ( $this->object_id_field_for_meta ) ) {
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-data.php]: I don't think this else needs to be nested.

File: includes/abstracts/abstract-wc-payment-token.php [hunk 3]
```diff
@@ -16,7 +16,7 @@
  * @category	Abstract Class
  * @author		WooThemes
  */
- abstract class WC_Payment_Token implements WC_Data {
+ abstract class WC_Payment_Token {
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-payment-token.php]: Not tested, but does this need to extend WC_Data?

File: includes/abstracts/abstract-wc-data.php (hunk 2 above)
Comment (mikejolley) [inline includes/abstracts/abstract-wc-data.php]: space after empty

File: includes/abstracts/abstract-wc-payment-token.php (hunk 3 above)
Comment (justinshreve) [inline includes/abstracts/abstract-wc-payment-token.php]: It should but will require a bit more changes then just extending the class to do it "properly" since the token stuff was written before the CRUD work started (for example, it has its own meta code). I'd like to do that in a separate PR, though.

This branch would also need to be merged in for 2.6 to get used by the token class in 2.6. Are we good with that?

File: includes/abstracts/abstract-wc-data.php [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+	 */
+	private function _get_db_info() {
+		global $wpdb;
+
+		$object_id_field = '';
+		$meta_id_field   = 'meta_id'; // for some reason users calls this umeta_id so we need to track this as well.
+		$table           = $wpdb->prefix;
+		// If we are dealing with a type of metadata that is not a core type, the table should be prefixed.
+		if ( ! in_array( $this->_meta_type, array( 'post', 'user', 'comment', 'term' ) ) ) {
+			$table .= 'woocommerce_';
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-data.php]: This statement could also be used for obiect_id_field if you wanted to cut down the size of this function. Each obiect_id_field is named {$this->_meta_type}_id to save on some code.

File: includes/abstracts/abstract-wc-data.php [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+		}
+		$this->read_meta_data();
+	}
+
+	/**
+	 * Table structure is slightly different between meta types, this function will return what we need to know.
+	 * @since 2.7.0
+	 * @return array Array elements: table, object_id_field, meta_id_field
+	 */
+	private function _get_db_info() {
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-data.php]: Should be this protected?

File: includes/abstracts/abstract-wc-payment-token.php (hunk 3 above)
Comment (mikejolley) [inline includes/abstracts/abstract-wc-payment-token.php]: Yeah, this won't break anything :) Maybe just need to tweak shipping zones which implemented WC_Data too.

File: includes/abstracts/abstract-wc-data.php [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+* Abstract WC Data Class
+*
+* Implemented by classes using the same CRUD(s) pattern.
+*
+* @version  2.7.0
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-data.php]: These versions need changing if merging to 2.6

**Comment by mikejolley** [comment]
:+1: Merging. I will work on shipping zones compatibility with this today.
